platforms/core-runtime/messaging/src/main/java/org/gradle/internal/remote/internal/inet/TcpIncomingConnector.java

```diff
@@ -109,13 +117,34 @@ public void run() {
                         InetSocketAddress remoteSocketAddress = (InetSocketAddress) socket.socket().getRemoteSocketAddress();
                         InetAddress remoteInetAddress = remoteSocketAddress.getAddress();
                         if (!allowRemote && !addressFactory.isCommunicationAddress(remoteInetAddress)) {
-                            LOGGER.error("Cannot accept connection from remote address {}.", remoteInetAddress);
+                            LOGGER.error("Cannot accept connection from remote address {}.", remoteSocketAddress);
                             socket.close();
                             continue;
                         }
-                        LOGGER.debug("Accepted connection from {} to {}.", socket.socket().getRemoteSocketAddress(), socket.socket().getLocalSocketAddress());
                         try {
                             SocketBlockingUtil.configureNonblocking(socket);
+                            ByteBuffer buffer = ByteBuffer.allocate(CONNECTION_PREAMBLE.length);
+                            long timeout = System.nanoTime() + TimeUnit.SECONDS.toNanos(acceptTimeoutSeconds);
+                            while (buffer.hasRemaining() && System.nanoTime() < timeout) {
+                                int read = socket.read(buffer);
+                                if (read == -1) {
+                                    break;
+                                }
+                                if (read == 0) {
+                                    Thread.sleep(10);
+                                }
+                            }
+                            if (!Arrays.equals(buffer.array(), CONNECTION_PREAMBLE)) {
+                                throw new IOException("Did not receive connection preamble within " + acceptTimeoutSeconds + "s");
+                            }
```

Comment:
If the buffer didn’t fill completely, buffer.array() may contain zeros (or leftover bytes), and the comparison will fail even if partial data was read. Note: I'm not familiar with this codebase so I may be wrong.

Suggestion:
```
    if (buffer.position() < CONNECTION_PREAMBLE.length ||
        !Arrays.equals(buffer.array(), CONNECTION_PREAMBLE)) {
        throw new IOException("Did not receive valid connection preamble within " + acceptTimeoutSeconds + "s");
    }
```

Reply:
>  the comparison will fail even if partial data was read

That's intentional. If we didn't receive the full string in the given time, then we want to fail.